Use Microsoft.Net.Compilers.Toolset in coreclr/src/Common projects

ViktorHofer opened the pull request:
- Remove dependencies.props favor of Versions.props
Versions.props is the Arcade entry point to manage version which are
being used throughout the repository.

- Use Arcade Sdk.targets in coreclr/src/Common
At the moment only Arcade Sdk.props are being imported in the common
managed projects. Changing that to make sure that the same compiler is
being used everywhere.

- Consolidate managed properties to OSGroup, ArchGroup and ConfigurationGroup. With the ongoing work done by @Anipik which will allow multi-targeting by OS this will be cleaned-up further.

- Packaging fixes to point to the right license and third party notices

- A lot of cleanup

cc @janvorli

ViktorHofer committed:
Remove dependencies.props favor of Versions.props

Versions.props is the Arcade entry point to manage version which are
being used throughout the repository.

ViktorHofer committed:
Use Arcade Sdk.targets in coreclr/src/Common

At the moment only Arcade Sdk.props are being imported in the common
managed projects. Changing that to make sure that the same compiler is
being used everywhere.

ViktorHofer committed:
Split restore and build targets in runtest.proj

Restore and Build targets shouldn't run in the same msbuild invocation as
that could cause issues with nuget auto-generated props/targets.

ViktorHofer committed:
Fix duplicate imports of liveBuild.targets

ViktorHofer committed:
Fix intermediate paths for test projects

Intermediate paths are encoded wrong so that
MSBuildProjectExtensionsPath which defaults to
BaseIntermediateOutputPath is set wrong and the in-built msbuild Exists
check fails. This is because of a mixture of different slashes.

ViktorHofer committed:
Various path fixes

safern commented on src/coreclr/tests/src/Common/Directory.Build.props:
```diff
@@ -0,0 +1,9 @@
+  <Import Project="..\Directory.Build.props" />
+
+  <PropertyGroup>
+    <UseLiveBuild>false</UseLiveBuild>
```
Why is this?

ViktorHofer replied:
Do we want to build every tool in our repo against the live ref assemblies of libraries? I spoke with @jkoritzinsky about this and he suggested to build these common tools and the generated xunit wrappers against the sdk refs. I can add a comment.

safern replied:
I thought that was the point of a combined repo, no? To build everything live?

safern replied:
Oh this is just a tool, not tests.

jkoritzinsky replied:
Yep. The assemblies in tests/src/Common other than CoreClrTestLibrary are tools to assist running the tests and are not part of the tests themselves.

safern replied:
Makes sense.

ViktorHofer committed:
MSBuild cleanup

ViktorHofer committed:
Update paths and common props

ViktorHofer committed:
Cleanup coreclr managed properties

ViktorHofer committed:
Split restore and build to fix design time imports

The restore and build target invocations need to happen in separate
msbuild task invocations and different global properties need to be
passed in to mitigate issues with caching the invocations. For more
context see https://github.com/microsoft/msbuild/issues/2811.

ViktorHofer committed:
Further path fixes

ViktorHofer committed:
Pre-Christmas changes

safern commented on src/coreclr/Directory.Build.props:
```diff
@@ -1,28 +1,168 @@
+    <ConfigurationGroup Condition="'$(__BuildType)' != ''">$(__BuildType)</ConfigurationGroup>
+    <ConfigurationGroup Condition="'$(ConfigurationGroup)' == '' or '$(ConfigurationGroup)' == 'debug'">Debug</ConfigurationGroup>
+    <ConfigurationGroup Condition="'$(ConfigurationGroup)' == 'release'">Release</ConfigurationGroup>
+    <ConfigurationGroup Condition="'$(ConfigurationGroup)' == 'checked'">Checked</ConfigurationGroup>
```
Can't we use ToUpper here?

ViktorHofer replied:
We could, but ToUpper converts all letters. We just want to upper-case the first one so I think this is the easiest solution in msbuild. Also this is just a move.

safern replied:
I thought there was an API to just upper case the first letter, but that's fine.

ViktorHofer replied:
I couldn't find it, does one exist?

safern replied:
Hmm it seems like there's not one for such operation, I thought there was in C#, apparently not.

safern commented on src/coreclr/Directory.Build.props:
```diff
@@ -1,28 +1,168 @@
+    </When>
+    <When Condition="'$(OSGroup)' == 'FreeBSD'">
+      <PropertyGroup>
+        <TargetsUnix>true</TargetsUnix>
```
This is all duplicated now with libraries OSGroup logic, can't we share that?

ViktorHofer replied:
We will be able to share it with @Anipik's changes but it isn't possible today.

safern commented on src/coreclr/Directory.Build.props:
```diff
@@ -1,28 +1,168 @@
+  -->
+  <PropertyGroup>
+    <MicrosoftNetCoreIlasmPackageRuntimeId Condition=" '$(TargetRid)' == 'linux-musl-x64' ">$(TargetRid)</MicrosoftNetCoreIlasmPackageRuntimeId>
+    <MicrosoftNetCoreIlasmPackageRuntimeId Condition=" '$(TargetRid)' == 'rhel.6-x64' ">$(TargetRid)</MicrosoftNetCoreIlasmPackageRuntimeId>
```
rhel.6-x64 is now longer supported. We no longer build it.

safern commented on src/coreclr/src/.nuget/packaging.props:
```diff
@@ -1,17 +1,13 @@
-    <PackageThirdPartyNoticesFile>$(CoreclrDir)/THIRD-PARTY-NOTICES.TXT</PackageThirdPartyNoticesFile>
-
+    <PackageDescriptionFile>$(MSBuildThisFileDirectory)descriptions.json</PackageDescriptionFile>
+    <PackageLicenseFile>$(RepoRoot)LICENSE.TXT</PackageLicenseFile>
```
Could `PackageThirdPartyNoticesFile` and `PackageLicenseFile` be shared with libraries packaging?

safern commented on src/coreclr/src/System.Private.CoreLib/System.Private.CoreLib.csproj:
```diff
@@ -6,14 +6,13 @@
     <EnsureRuntimePackageDependencies>false</EnsureRuntimePackageDependencies>
-    <TargetFramework>netcoreapp2.1</TargetFramework>
+    <TargetFramework>$(NetCoreAppCurrent)</TargetFramework>
+    <TargetFramework Condition="'$(BuildingInsideVisualStudio)' == 'true'">netcoreapp3.0</TargetFramework>
```
Why can't we build for nca5.0 in visual studio?

ViktorHofer replied:
Unsure, see https://github.com/dotnet/runtime/pull/584 for discussion.

safern commented on src/coreclr/src/gc/sample/GCSample.vcxproj:
```diff
@@ -17,13 +17,13 @@
   </PropertyGroup>
   <Import Project="$(VCTargetsPath)\Microsoft.Cpp.Default.props" />
-  <PropertyGroup Condition="'$(Configuration)|$(Platform)'=='Debug|Win32'" Label="Configuration">
+  <PropertyGroup Condition="'$(ConfigurationGroup)|$(ArchGroup)'=='Debug|Win32'" Label="Configuration">
```
We could probably create a property for this value and use that for the condition. Also, spaces in between `'` and `==`.

ViktorHofer replied:
I don't think it's worth it. I just replaced the property name here.

safern commented on src/coreclr/tests/src/dirs.proj:
```diff
@@ -45,6 +45,10 @@
     </ItemGroup>
 
+    <PropertyGroup>
+      <__TestGroupToBuild Condition="'$(__TestGroupToBuild)' == ''">1</__TestGroupToBuild>
```
Where was this defined before? Or is it to avoid having to pass it down?

ViktorHofer replied:
right, otherwise msbuild would complain. Currently this is always invoked from scripts but we will change this in the future. So this is just to unblock invoking this project directly in the meantime.

safern commented on src/installer/Directory.Build.props:
```diff
@@ -39,8 +39,6 @@
 
   <!-- Platform detection -->
   <PropertyGroup>
-    <RunningOnUnix Condition="'$(OS)'!='Windows_NT'">true</RunningOnUnix>
```
Was this not used at all?

ViktorHofer replied:
I just moved this into the root.

ViktorHofer committed:
Share packaging props

ViktorHofer committed:
Remove rhel6 instance

ViktorHofer committed:
Use PackageLicenseExpression over PackageLicenseFile

ViktorHofer committed:
More packaging cleanup

ViktorHofer committed:
Remove default set properties

ViktorHofer committed:
Revert some platform property changes

ViktorHofer committed:
Remove ArchGroup property in favor of Platform

ViktorHofer committed:
Remove ConfigurationGroup in favor of Configuration

ViktorHofer committed:
Cleanup

stephentoub commented:
@ViktorHofer, what's the state of this?  Thanks.

stephentoub commented:
@ViktorHofer, given this has been open for two months, hasn't been touched in a month, has a bunch of conflicts, and there's been no response in over a week about status, I'm going to close this.  It can obviously be re-opened when you're ready to proceed with it.  Thanks.

ViktorHofer commented:
I've been working on this over the weekend but am still not in a state to push an update. Closing is fine, I'll reopen when this is ready again.

stephentoub commented:
Thanks.
